Polkadot nominator and validator guides (#467)
w3f/polkadot-wiki: docs/maintain-guides-how-to-validate-polkadot.md

---

> id: maintain-guides-how-to-validate-kusama

Reviewer: ```suggestion
id: maintain-guides-how-to-validate-polkadot
```

> title: Run a Validator (Kusama)

Reviewer: ```suggestion
title: Run a Validator (Polkadot)
```

very rewarding, knowing that you contribute to the security of a decentralized network while growing
your stash.

Since security is so important to running a successful validator, you should take a look at the
[secure validator](maintain-guides-secure-validator) information to make you understand the factors
to consider when constructing your infrastructure. The Web3 Foundation also maintains a
[reference implementation for a secure validator set-up](https://github.com/w3f/polkadot-secure-validator)
that you can use by deploying yourself (video walkthrough is available
[here](https://www.youtube.com/watch?v=tTn8P6t7JYc)). As you progress in your journey as a
validator, you will likely want to use this repository as a _starting point_ for your own
modifications and customizations.

If you need help, please reach out on the
[Polkadot Validator Longue](https://matrix.to/#/!NZrbtteFeqYKCUGQtr:matrix.parity.io?via=matrix.parity.io&via=matrix.org&via=web3.foundation)
on Riot. The team and other validators are there to help answer questions and provide tips from
experience. If you have a more significant proposal, you can write it on the

> [Web3 Foundation forum](https://forum.web3.foundation/).

Reviewer: I heard word of deprecating the discourse forums - nothing really happens here. Is this worth mentioning?

Author: It doesn't get much attention. Probably opening up an issue on one of the relevant repositories on GitHub gets more visibility.

brew install cmake pkg-config openssl git llvm
```

### Install & Configure Network Time Protocol (NTP) Client

[NTP](https://en.wikipedia.org/wiki/Network_Time_Protocol) is a networking protocol designed to
synchronize the clocks of computers over a network. NTP allows you to synchronize the clocks of all
the systems within the network. Currently it is required that validators' local clocks stay
reasonably in sync, so you should be running NTP or a similar service. You can check whether you
have the NTP client by running:

_If you are using Ubuntu 18.04 / 19.04, NTP Client should be installed by default._

```sh
timedatectl
```

If NTP is installed and running, you should see `System clock synchronized: yes` (or a similar
message). If you do not see it, you can install it by executing:

```sh
sudo apt-get install ntp
```

ntpd will be started automatically after install. You can query ntpd for status information to
verify that everything is working:

```sh
sudo ntpq -p
```

### Building and Installing the `polkadot` Binary

You will need to build the `polkadot` binary from the
[paritytech/polkadot](https://github.com/paritytech/polkadot) repository on GitHub using the source

> code available in the **v0.7** branch.

Reviewer: ```suggestion
code available in the **v0.8** branch.
```

> You should generally use the latest **0.7.x** tag. At the time of writing, this was **0.7.28**, but

Reviewer: ```suggestion
You should generally use the latest **0.8.x** tag. At the time of writing, this was **0.8.0**, but
```

> you should review the output from the "git tag" command (`git tag | grep "$v\0\.7"`) to see a list

Reviewer: ```suggestion
you should review the output from the "git tag" command (`git tag | grep "$v\0\.8"`) to see a list
```

> of all the potential 0.7 releases. You should replace `v0.7.28` with the latest build (i.e., the

Reviewer: ```suggestion
of all the potential 0.8 releases. You should replace `v0.8.0` with the latest build (i.e., the
```

highest number). You can also find the latest Kusama version on the
[release](https://github.com/paritytech/polkadot/releases) tab.

> Note: If you prefer to use SSH rather than HTTPS, you can replace the first line of the below with
> `git clone git@github.com:paritytech/polkadot.git`.

```sh
git clone https://github.com/paritytech/polkadot.git
cd polkadot

> git tag | grep "$v\0\.7"

Reviewer: ```suggestion
git tag | grep "$v\0\.8"
```

> git checkout v0.7.28

Reviewer: ```suggestion
git checkout v0.8.0
```

./scripts/init.sh
cargo build --release

> ```

Reviewer: This will change, but not really sure what it will look like in the meantime.....

This step will take a while (generally 15 - 30 minutes, depending on your hardware).

If you are interested in generating keys locally, you can also install `subkey` from the same
directory. You may then take the generated `subkey` executable and transfer it to an air-gapped
machine for extra security.

```sh
cargo install --force --git https://github.com/paritytech/substrate subkey
```

### Synchronize Chain Data

> > **Note:** Validators must sync their nodes in archive mode to avoid being slashed. If you've

Reviewer: I think this can change a bit - validators won't be slashed, but their nodes just won't start (which can mean slashing).

**Note:** If a Validator syncs as a full node, they must sync their nodes in archive mode, otherwise the node will throw an error when on start.

Reviewer: Actually you do not need to run in archive mode, you can run in unsafe pruning mode with the following flags:

`--unsafe-pruning --pruning <pruning number>`

Where pruning number can be something like 16384

> already synced the chain, you must first remove the database with `polkadot purge-chain` and then
> ensure that you run Polkadot with the `--pruning=archive` option.

You can begin syncing your node by running the following command:

> 

Reviewer: There should be something here on rocksdb vs paritydb - right now people should use rocksdb, but paritydb will be faster and eventually we will switch to that. If you use one vs the other it will require a resync.

```sh
./target/release/polkadot --pruning=archive
```

if you do not want to start in validator mode right away.

**Note:** The `--pruning=archive` flag is implied by the `--validator` and `--sentry` flags, so it
is only required explicitly if you start your node without one of these two options. If you do not
set your pruning to archive node, even when not running in validator and sentry mode, you will need
to re-sync your database when you switch.

Depending on the size of the chain when you do this, this step may take anywhere from a few minutes
to a few hours.

If you are interested in determining how much longer you have to go, your server logs (printed to
STDOUT from the `polkadot` process) will tell you the latest block your node has processed and
verified. You can then compare that to the current highest block via
[Telemetry](https://telemetry.polkadot.io/#list/Kusama) or the
[PolkadotJS Block Explorer](https://polkadot.js.org/apps/#/explorer).

> > **Note:** If you do not already have KSM, this is as far as you will be able to go until the end

Reviewer: ```suggestion
> **Note:** If you do not already have DOTs, this is as far as you will be able to go until the end
```

> of the soft launch period. You can still run a node, but you will need to have a minimal amount of

> > KSM to continue, as balance transfers are disabled during the soft launch. Please keep in mind

Reviewer: ```suggestion
> DOTs to continue, as balance transfers are disabled during the soft launch. Please keep in mind
```

> > that even for those with KSM, they will only be indicating their _intent_ to validate; they will

Reviewer: ```suggestion
> that even for those with DOTs, they will only be indicating their _intent_ to validate; they will
```

> also not be able to run a validator until the NPoS phase starts.

> ## Bond KSM

Reviewer: ```suggestion
## Bond DOTs
```

> 

Reviewer: Add a note here that until transfers are enabled, you _must_ use the same account if you did not get DOT injections done.

It is highly recommended that you make your controller and stash accounts be two separate accounts.
For this, you will create two accounts and make sure each of them have at least enough funds to pay
the fees for making transactions. Keep most of your funds in the stash account since it is meant to
be the custodian of your staking funds.

> Make sure not to bond all your KSM balance since you will be unable to pay transaction fees from

Reviewer: ```suggestion
Make sure not to bond all your DOT balance since you will be unable to pay transaction fees from
```

your bonded balance.

It is now time to set up our validator. We will do the following:

> - Bond the KSM of the Stash account. These KSM will be put at stake for the security of the network

Reviewer: ```suggestion
- Bond the DOTs of the Stash account. These DOTs will be put at stake for the security of the network
```

  and can be slashed.
- Select the Controller. This is the account that will decide when to start or stop validating.

First, go to the [Staking](https://polkadot.js.org/apps/#/staking/actions) section. Click on
"Account Actions", and then the "New stake" button.

![dashboard bonding](assets/guides/how-to-validate/polkadot-dashboard-bonding.jpg)

> - **Stash account** - Select your Stash account. In this example, we will bond 100 milliKSMs - make

Reviewer: ```suggestion
- **Stash account** - Select your Stash account. In this example, we will bond 100 milliDOTs - make
```

  sure that your Stash account contains _at least_ this much. You can, of course, stake more than
  this.
- **Controller account** - Select the Controller account created earlier. This account will also

>   need a small amount of KSM in order to start and stop validating.

Reviewer: ```suggestion
  need a small amount of DOTs in order to start and stop validating.
```

> - **Value bonded** - How much KSM from the Stash account you want to bond/stake. Note that you do

Reviewer: ```suggestion
- **Value bonded** - How much DOts from the Stash account you want to bond/stake. Note that you do
```

>   not need to bond all of the KSM in that account. Also note that you can always bond _more_ KSM

Reviewer: ```suggestion
  not need to bond all of the DOTs in that account. Also note that you can always bond _more_ DOTs
```

Once everything is filled in properly, click `Bond` and sign the transaction with your Stash
account.

After a few seconds, you should see an "ExtrinsicSuccess" message. You should now see a new card
with all your accounts (note: you may need to refresh the screen). The bonded amount on the right
corresponds to the funds bonded by the Stash account.

## Set Session Keys

> **Note:** The session keys are consensus critical, so if you are not sure if your node has the
> current session keys that you made the `setKeys` transaction then you can use one of the two
> available RPC methods to query your node:
> [hasKey](https://polkadot.js.org/api/substrate/rpc.html#haskey-publickey-bytes-keytype-text-bool)
> to check for a specific key or
> [hasSessionKeys](https://polkadot.js.org/api/substrate/rpc.html#hassessionkeys-sessionkeys-bytes-bool)
> to check the full session key public key string.

Once your node is fully synced, stop the process by pressing Ctrl-C. At your terminal prompt, you
will now start running the node in validator mode with a flag allowing unsafe RPC calls, needed for
some advanced operations.

```sh

> ./target/release/polkadot --validator --name "name on telemetry" --unsafe-rpc-expose

Reviewer: ```suggestion
./target/release/polkadot --validator --name "name on telemetry"
```

You shouldn't actually need the --unsafe-rpc-expose flag anymore, the addition of this initially was a bug.

```

You can give your validator any name that you like, but note that others will be able to see it, and
it will be included in the list of all servers using the same telemetry server. Since numerous
people are using telemetry, it is recommended that you choose something likely to be unique. Note
that the `--rpc-methods=Unsafe` flag is needed in order to set the session key as shown below. Note

> this flag only exposes RPC methods locally.

Reviewer: Can remove the note on the unsafe rpc methods.

Reviewer: With the V0.8.0 release, the necessity of this was removed.

would effect the local keystore hosted on a _public node_ and you want to make sure you are
interacting with the keystore for _your node_.

Once ensuring that you have connected to your node, the easiest way to set session keys for your
node is by calling the `author_rotateKeys` RPC request to create new keys in your validator's
keystore. Navigate to Toolbox tab and select RPC Calls then select the author > rotateKeys() option
and remember to save the output that you get back for a later step.

![Explorer RPC call](assets/guides/how-to-validate/polkadot-explorer-rotatekeys-rpc.jpg)

#### Option 2: CLI

If you are on a remote server, it is easier to run this command on the same machine (while the node
is running with the default HTTP RPC port configured):

```sh
curl -H "Content-Type: application/json" -d '{"id":1, "jsonrpc":"2.0", "method": "author_rotateKeys", "params":[]}' http://localhost:9933
```

The output will have a hex-encoded "result" field. The result is the concatenation of the four
public keys. Save this result for a later step.

You can restart your node at this point, omitting the `--unsafe-rpc-expose` flag as it is no longer

> needed.

Reviewer: Remove mention of unsafe rpc.

### Submitting the `setKeys` Transaction

You need to tell the chain your Session keys by signing and submitting an extrinsic. This is what
associates your validator with your Controller account.

Go to [Staking > Account Actions](https://polkadot.js.org/apps/#/staking/actions), and click "Set
Session Key" on the bonding account you generated earlier. Enter the output from `author_rotateKeys`
in the field and click "Set Session Key".

![staking-change-session](assets/guides/how-to-validate/set-session-key-1.jpg)
![staking-session-result](assets/guides/how-to-validate/set-session-key-2.jpg)

Submit this extrinsic and you are now ready to start validating.

## Validate

To verify that your node is live and synchronized, head to

> [Telemetry](https://telemetry.polkadot.io/#/Kusama) and find your node. Note that this will show all

Reviewer: ```suggestion
[Telemetry](https://telemetry.polkadot.io/#list/Polkadot%20CC1) and find your node. Note that this will show all
```

> [Telemetry](https://telemetry.polkadot.io/#/Kusama) and find your node. Note that this will show all

Reviewer: Link should be polkadot telemetry

> nodes on the Kusama network, which is why it is important to select a unique name!

Reviewer: ```suggestion
nodes on the Polkadot network, which is why it is important to select a unique name!
```

  nominators.

Click "Validate".

If you go to the "Staking" tab, you will see a list of active validators currently running on the
network. At the top of the page, it shows how many validator slots are available as well as how many
nodes have signaled their intention to be a validator. You can also go to the "Waiting" tab to
double check to see whether your node is listed there.

![staking queue](assets/guides/how-to-validate/polkadot-dashboard-staking.jpg)

The validator set is refreshed every era. In the next era, if there is a slot available and your
node is selected to join the validator set, your node will become an active validator. Until then,
it will remain in the _waiting_ queue. If your validator is not selected to become part of the
validator set, it will remain in the _waiting_ queue until it is. There is no need to re-start if
you are not selected for the validator set in a particular era. However, it may be necessary to
increase the number of KSMs staked or seek out nominators for your validator in order to join the
validator set.

**Congratulations!** If you have followed all of these steps, and been selected to be a part of the

> validator set, you are now running a Kusama validator! If you need help, reach out on the

Reviewer: ```suggestion
validator set, you are now running a Polkadot validator! If you need help, reach out on the
```

> [Kusama forum](https://forum.kusama.network/) or in the

Reviewer: ```suggestion
[Web3 Foundation forum](https://forum.web3.foundation/) or in the
```

> [Kusama Validator chat](https://riot.im/app/#/room/#KusamaValidatorLounge:polkadot.builders).

Reviewer: ```suggestion
[Polkadot Validator chat](https://matrix.to/#/!NZrbtteFeqYKCUGQtr:matrix.parity.io?via=matrix.parity.io&via=matrix.org&via=web3.foundation).
```

> [Kusama Validator chat](https://riot.im/app/#/room/#KusamaValidatorLounge:polkadot.builders).

Reviewer: Replace with Polkadot links.

## FAQ

### Why am I unable to synchronize the chain with 0 peers?

![zero-peer](assets/guides/how-to-validate/polkadot-zero-peer.jpg)

Make sure to enable `30333` libp2p port. Eventually, it will take a little bit of time to discover
other peers over the network.

### How do I clear all my chain data?

```sh
./target/release/polkadot purge-chain
```

## VPS List

- [OVH](https://www.ovh.com.au/)
- [Digital Ocean](https://www.digitalocean.com/)
- [Vultr](https://www.vultr.com/)
- [Linode](https://www.linode.com/)
- [Contabo](https://contabo.com/)
- [Scaleway](https://www.scaleway.com/)

## Using Docker

If you have Docker installed, you can use it to start your validator node without needing to build
the binary. You can do this with a simple one line command:

```sh

> $ docker run parity/polkadot:v0.7.28 --validator --name "name on telemetry"

Reviewer: ```suggestion
$ docker run parity/polkadot:latest --validator --name "name on telemetry"
```

if you do not want to start in validator mode right away.

The `--pruning=archive` flag is implied by the `--validator` and `--sentry` flags, so it is only
required explicitly if you start your node without one of these two options. If you do not set your
pruning to archive node, even when not running in validator and sentry mode, you will need to
re-sync your database when you switch.

> > **Note:** Validators should sync using the RocksDb backend by passing the `--databse RocksDb`

Reviewer: ```suggestion
> **Note:** Validators should sync using the RocksDb backend by passing the `--database RocksDb`
```
